src/app/security-group/components/sg-rule-addition-form/sg-rule-addition-form.component.ts

```diff
@@ -0,0 +1,414 @@
+    let translateToken = 'SECURITY_GROUP_PAGE.RULES.';
+    let param: { value?: string } = {};
+
+    if (port.hasError('required')) {
+      translateToken += 'FILED_REQUIRED';
+    } else if (port.hasError('min')) {
+      translateToken += 'START_PORT_SHOULD_BE_GREATER_THAN';
+      param = { value: port.errors.min.min }
+    } else if (port.hasError('max')) {
+      translateToken += 'START_PORT_SHOULD_BE_LESS_THAN';
+      param = { value: port.errors.max.max }
+    } else if (port.hasError('startPortValidator')) {
+      translateToken += 'START_PORT_SHOULD_BE_LESS_THAN_END_PORT';
+    }
+
+    return this.translateService.instant(translateToken, param);
+  }
+
+  public getEndPortErrorMessage() {
+    const port = this.portsForm.get('endPort');
+    let translateToken = 'SECURITY_GROUP_PAGE.RULES.';
+    let param: { value?: string } = {};
+
+    if (port.hasError('required')) {
+      translateToken += 'FILED_REQUIRED';
+    } else if (port.hasError('min')) {
+      translateToken += 'END_PORT_SHOULD_BE_GREATER_THAN';
+      param = { value: port.errors.min.min }
+    } else if (port.hasError('max')) {
+      translateToken += 'END_PORT_SHOULD_BE_LESS_THAN';
+      param = { value: port.errors.max.max }
+    } else if (port.hasError('endPortValidator')) {
+      translateToken += 'END_PORT_SHOULD_BE_GREATER_THAN_START_PORT';
+    }
+
+    return this.translateService.instant(translateToken, param);
+  }
+
+  private getFormData(): FirewallRule {
+    const formModel = this.ruleForm.value;
+    const rule: FirewallRule = {
+      type: formModel.type,
+      protocol: formModel.protocol,
+      cidr: formModel.cidr
+    };
+    if (rule.protocol === NetworkProtocol.ICMP) {
+      const icmpModel = this.icmpForm.value;
+      rule.icmpType = icmpModel.icmpType;
+      rule.icmpCode = icmpModel.icmpCode;
+    } else {
+      const portsModel = this.portsForm.value;
+      rule.startPort = portsModel.startPort;
+      rule.endPort = portsModel.endPort;
+    }
+    return rule;
+  }
+
+  private resetForm() {
+    const paramsForm = this.ruleForm.get('protocol').value === NetworkProtocol.ICMP ? this.icmpForm : this.portsForm;
+    const formState = {
+      type: this.ruleForm.get('type').value,
+      protocol: this.ruleForm.get('protocol').value,
+      cidr: this.ruleForm.get('cidr').value,
+      params: paramsForm
+    };
+    this.mainForm.resetForm();
+    this.ruleForm.reset(formState);
+  }
+
+  private setPortFormValidators() {
+    const portCommonValidators = [
+      Validators.required,
+      Validators.min(this.minPortNumber),
+      Validators.max(this.maxPortNumber)
+    ];
+    const startPort = this.portsForm.get('startPort');
+    const endPort = this.portsForm.get('endPort');
+    startPort.setValidators([...portCommonValidators, startPortValidator(endPort)]);
+    endPort.setValidators([...portCommonValidators, endPortValidator(startPort)]);
+  }
+
+  private onProtocolChange() {
```

Comment:
can we add a handler for an event emitter of the protocol field?
same for other subscriptions

Reply:
Yes, we can, but the idea of reactive forms to put eject logic from a template and place it to a component. Also using this way I receive an observable and I can transform it before use. It will be much clearer to use this approach.